ai/core/src/unitycatalog/ai/core/executor/local.py

```diff
@@ -0,0 +1,177 @@
+        memory_limit: Maximum memory in MB.
+    """
+    try:
+        resource.setrlimit(resource.RLIMIT_CPU, (cpu_time_limit, cpu_time_limit))
+    except Exception as e:
+        _logger.info("Warning: unable to set RLIMIT_CPU: %s", e)
+
+    try:
+        memory_bytes = memory_limit * MB_CONVERSION
+        resource.setrlimit(resource.RLIMIT_AS, (memory_bytes, memory_bytes))
+    except Exception as e:
+        _logger.info("Warning: unable to set RLIMIT_AS: %s", e)
+
+
+def _disable_unwanted_imports():
+    """
+    Override the built-in __import__ and open function to block
+    potentially dangerous modules and file access.
+    """
+    original_import = builtins.__import__
+
+    def restricted_import(name, globals=None, locals=None, fromlist=(), level=0):
+        if name in DISALLOWED_MODULES:
+            raise ImportError(generate_import_disallowed_message(name))
+        return original_import(name, globals, locals, fromlist, level)
+
+    builtins.__import__ = restricted_import
+
+    def disabled_open(*args, **kwargs):
+        raise ImportError(OPEN_DISALLOWED_MESSAGE)
+
+    builtins.open = disabled_open
+
+
+def _sandboxed_wrapper(
+    q, func: Callable[..., Any], params: dict[str, Any], cpu_time_limit: int, memory_limit: int
+):
+    """
+    Execute the provided callable in a sandboxed environment.
+
+    Applies resource limits and restricts dangerous module imports.
+    The function `func` is called with keyword arguments from `params`.
+    If the function returns a coroutine, it is executed via asyncio.run().
+    The result or any exception's full stack trace is placed on a queue.
+    """
+    try:
+        _limit_resources(cpu_time_limit, memory_limit)
+        _disable_unwanted_imports()
+        result = func(**params)
+        if asyncio.iscoroutine(result):
+            result = asyncio.run(result)
+        parsed_result = result if result is not None else NO_OUTPUT_MESSAGE
+        q.put((True, parsed_result))
+    except Exception:
+        tb = traceback.format_exc()
+        q.put((False, tb))
+
+
+async def run_in_sandbox_async(
+    func: Callable[..., Any], params: dict[str, Any]
+) -> tuple[bool, Any]:
+    """
+    Executes a Python callable in a sandboxed subprocess using multiprocessing.
+    Specific core Python modules are restricted to prevent unwanted behavior.
+    The function is executed with a timeout and resource limits for CPU and memory.
+    The function's result or any exception's full stack trace is returned.
+
+    Parameters:
+        func: The callable to execute.
+        params: A dictionary of keyword arguments to pass to the function.
+
+    Returns:
+        A tuple (success, result). If success is False, result contains a descriptive
+        error message or the error stack trace.
+    """
+    cpu_time_limit = EXECUTOR_MAX_CPU_TIME_LIMIT.get()
+    memory_limit = EXECUTOR_MAX_MEMORY_LIMIT.get()
+    timeout = EXECUTOR_TIMEOUT.get()
+
+    ctx = get_context("fork")
```

Comment:
Would `fork` be dangerous on macOS? ref: https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods